feat: add new formatter log output for github

fzipi opened the pull request:
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

Modify logger to output in a format parseable by Github's runners.

Fixes #56.

Pending:
- [x] add tests on output

fzipi committed:
feat: add new formatter log output for github

Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

fzipi committed:
feat: print using levels

Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

fzipi commented on chore/update_copyright.go:
```diff
@@ -68,12 +69,17 @@ func updateRules(version string, year string, contents []byte) ([]byte, error) {
 	writer := bufio.NewWriter(output)
 	replaceVersion := fmt.Sprintf("${1}%s", version)
+	// only keep numbers from the version
+	semanticVersion := regexp.MustCompile(`(\d)\.(\d)\.(\d)*`)
```
Probably need to add release candidate support `(-rc)?`

fzipi replied:
Changed that name, we didn't' needed an actual "semantic version" there.

fzipi commented:
Moved commit 9ea579cf0df5c76cd777e1a0025ec4dbdddad256 to #122.

fzipi committed:
fix: use logger only for github messages

Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

fzipi commented:
@theseion Looks like test is failing (cmd/regex_compare_test.go:151) because the error was not triggered per `if rootValues.output == gitHub`... so probably we want to change the test?

fzipi commented on cmd/regex_compare.go:
```diff
@@ -148,9 +148,8 @@ func performCompare(processAll bool, ctx *processors.Context) error {
-		if failed && rootValues.output == gitHub {
-			fmt.Println("::error::All rules need to be up to date.",
-				"Please run `crs-toolchain regex update --all`")
+		if failed {
+			logger.Error().Msg("All rules need to be up to date. Please run `crs-toolchain regex update --all`")
```
This change was shadowing a check failure, because the test wasn't being executed in GH mode.

theseion commented on logger/logger_test.go:
```diff
@@ -0,0 +1,124 @@
+var testGithub = []struct {
+	name    string
+	text    string
+	logType zerolog.Level
```
```suggestion
	logLevel zerolog.Level
```

theseion commented on logger/logger_test.go:
```diff
@@ -0,0 +1,124 @@
+	}
+}
+
+//s.log = log.Output(zerolog.ConsoleWriter{Out: os.Stderr, TimeFormat: "03:04:05"}).With().Caller().Logger()
```
```suggestion
```

theseion commented on logger/logger_test.go:
```diff
@@ -0,0 +1,124 @@
+//s.log = log.Output(zerolog.ConsoleWriter{Out: os.Stderr, TimeFormat: "03:04:05"}).With().Caller().Logger()
+
+func (s *loggerTestSuite) TestSetGithubOutput() {
+	// send logs to buffer
```
```suggestion
```

theseion commented on logger/logger_test.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+func (s *loggerTestSuite) TestConsoleOutput() {
+	s.logger = s.logger.Output(zerolog.ConsoleWriter{Out: s.out, NoColor: true, TimeFormat: "03:04:05"})
```
Instead of configuring the console output here explicitly for the test, wouldn't it be better to have an additional choice to generate JSON output? Just like you did now for GH.

fzipi replied:
Did you mean console output? JSON is the default. Do you want something like `SetConsoleOutput`?

theseion replied:
Zerolog's default is JSON, but our global logger is configured to use the console writer. So for users, currently, there's only the option of either console or GH optimised output. I don't see the point in testing JSON output if we don't offer it to users.

theseion reviewed (changes requested):
❤️

theseion commented on logger/logger_test.go:
```diff
@@ -0,0 +1,124 @@
+		text:    "this is a warning message",
+		logType: zerolog.WarnLevel,
+		want:    "::warn ::this is a warning message",
+	},
```
Add a test for error.

fzipi committed:
tests: added test for error/panic

Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

theseion commented:
@fzipi I've unresolved three comments, as the suggestions were not applied when I think they should have been.

theseion committed:
Merge branch 'main' into add-github-zerolog-output

fzipi commented:
@copilot Can you address theseion's open comments?

Copilot committed:
refactor(test): rename logType to logLevel in test structs, remove stale comments

Agent-Logs-Url: https://github.com/coreruleset/crs-toolchain/sessions/1505ba03-1173-4575-bc23-2783510abecc

Co-authored-by: fzipi <3012076+fzipi@users.noreply.github.com>

Copilot committed:
No new changes

Agent-Logs-Url: https://github.com/coreruleset/crs-toolchain/sessions/1505ba03-1173-4575-bc23-2783510abecc

Co-authored-by: fzipi <3012076+fzipi@users.noreply.github.com>
